# Data validation between the data set in BigQuery and GraphDB

The objectives are to:
1. Ensure complete upload of data to BigQuery and GraphDB.
2. Check for data qualtiy issues.
3. Gain insights on the data.
4. Engineer features.

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [9]:
# Mount to Google Drive to save results
from google.colab import drive
drive.mount('/content/drive')
%cd /content/drive/MyDrive/MSc/2020-21/Research\ Project/Colab/
%ls

Mounted at /content/drive
/content/drive/MyDrive/MSc/2020-21/Research Project/Colab


In [10]:
# Connect to GCP Bucket
from google.colab import auth
auth.authenticate_user()

In [11]:
# Set GCP project ID and region to Europe West 2 - London
PROJECT = 'detect-fake-news-313201'
!gcloud config set project $PROJECT
REGION = 'europe-west2'
CLUSTER = '{}-cluster'.format(PROJECT)
!gcloud config set compute/region $REGION
!gcloud config set dataproc/region $REGION

!gcloud config list # show some information

Updated property [core/project].
Updated property [compute/region].
Updated property [dataproc/region].
[component_manager]
disable_update_check = True
[compute]
gce_metadata_read_timeout_sec = 0
region = europe-west2
[core]
account = aaron.altrock@gmail.com
project = detect-fake-news-313201
[dataproc]
region = europe-west2

Your active configuration is: [default]


## Check the number of files in successive GCP cloud storage buckets

In [12]:
# Count the number of cleaned JSON files from the end of stage 1 in the pipeline
!gsutil ls -l gs://fake_news_cleaned_json/*.json | wc -l

59733


In [13]:
# Count the number of parsed JSON and TTL files into triples at the end of stage 2 in the pipeline
!gsutil ls -l gs://fake_news_ttl_json/*.ttl | wc -l
!gsutil ls -l gs://fake_news_ttl_json/*.json | wc -l

27590
27590


The variance between the 59,733 cleaned files to 27,590 turtle documents would suggest this is due to the raw data containing duplicating records for the same news web page, when the turtles are indexed by the hash value of the URLs and therefore would overwrite leading to small number of samples.

In [14]:
# Based on https://cloud.google.com/bigquery/docs/quickstarts/quickstart-client-libraries#bigquery_simple_app_client-python
from google.cloud import bigquery
client = bigquery.Client(PROJECT)


## Profile the data in its original form held in BigQuery

In [15]:
# BIgQuery data row count
query_job = client.query(
    """
    SELECT COUNT(*) AS POPULATION_COUNT
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

res_df

POPULATION_COUNT
0             27589

Therefore deviation by one record compared to the number of files in `gs://fake_news_ttl_json`.

In [16]:
# BIgQuery data row count
query_job = client.query(
    """
    WITH URL_LIST AS (
      SELECT 
      URL
      , COUNT(*) AS URL_COUNT
      FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
      GROUP BY URL
    )
    SELECT * FROM URL_LIST WHERE URL_COUNT > 1
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

res_df

Empty DataFrame
Columns: [URL, URL_COUNT]
Index: []

Therefore no samples found to have duplicating URL in the BigQuery table, and all articles have unique URLs.

In [17]:
# BIgQuery data preview
query_job = client.query(
    """
    SELECT *
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    LIMIT 10
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

res_df

url   domain  \
0  http://awm.com/woman-adopts-rescue-dog-starts-...  awm.com   
1  http://awm.com/grand-daughter-sings-favorite-s...  awm.com   
2  http://awm.com/first-class-passenger-sees-a-so...  awm.com   
3  http://awm.com/model-who-boasts-she-transforme...  awm.com   
4  http://awm.com/dad-puts-his-son-into-the-game-...  awm.com   
5  http://awm.com/little-girls-reaction-to-shooti...  awm.com   
6  http://awm.com/out-of-all-the-holiday-recipes-...  awm.com   
7  http://awm.com/a-starbucks-barista-couldnt-fin...  awm.com   
8  http://awm.com/the-easily-offended-whine-when-...  awm.com   
9  http://awm.com/during-heated-love-making-wife-...  awm.com   

                               domain_hash  \
0  domain_9725d0802cf38288710d8bff8f64dcba   
1  domain_9725d0802cf38288710d8bff8f64dcba   
2  domain_9725d0802cf38288710d8bff8f64dcba   
3  domain_9725d0802cf38288710d8bff8f64dcba   
4  domain_9725d0802cf38288710d8bff8f64dcba   
5  domain_9725d0802cf38288710d8bff8f64dcba   
6  domain_9725d0802cf38288710d8bff8f64dcba   
7  domain_9725d0802cf38288710d8bff8f64dcba   
8  domain_9725d0802cf38288710d8bff8f64dcba   
9  domain_9725d0802cf38288710d8bff8f64dcba   

                               title_hash  \
0  title_761e82d05283bff6f1832c7164626e26   
1  title_2a30a9b6a5a75250a7317c60c0f871e6   
2  title_b89b152a0c80235c4e6174a96e98f166   
3  title_9868eac5482fdbfebf7b94494bd078ec   
4  title_08db9d62bccf3a51cb7ded0fdf666568   
5  title_0ef200f6649ab2907625d4db1f796c87   
6  title_0165dfcd088b4df972d8b23ee45198e0   
7  title_55c16d853bfc26cc4697fe2158f2e273   
8  title_ab7f6ea47a85aeacab878345864501df   
9  title_3ed2e9ed1879767b09e706f2a7f5ced5   

                                               title  \
0  Woman Adopts Rescue Dog, Starts Noticing Foul ...   
1  Grand Daughter Sings Favorite Song to Grandma ...   
2  First Class Passenger Sees A Soldier Walking U...   
3  Model Who Boasts She Transformed Into An Afric...   
4  Dad Puts His Son Into The Game Days After Beat...   
5  Little Girl’s Reaction To Shooting Her First D...   
6  Out Of All The Holiday Recipes I’ve Tried, His...   
7  A Starbucks Barista Couldn’t Find A Babysitter...   
8  The Easily Offended Whine When Kelly Clarkson ...   
9  During Heated Love Making, Wife Accidentally B...   

                               body_hash  \
0  body_072ebc1af96077202eb2ad843b6fe4cd   
1  body_3e0c1daeb28fa250067a74240a5af93d   
2  body_6953f3cbe02f2bdf2083937b1c8f9b7f   
3  body_d9e3185ae65d5b97185f08d85c31abec   
4  body_fcb8636125a87f14428856aff72e307f   
5  body_c8d4436c055c1516d67af592b850a17f   
6  body_7a5cb6d058bda9afc409df365438cc7f   
7  body_b2277eca06c8bd84aba7a6512dcbb5e6   
8  body_ddf7af702f3aef83dcb93cc5717ad34f   
9  body_f62f488c90df847302cd010ff91e016b   

                                                body       label  \
0  Kelly Benzel was looking for a new best friend...  unreliable   
1  Dementia is one of the scariest and heartbreak...  unreliable   
2  Air travel can be a tiring and trying experien...  unreliable   
3  Martina Big, a 28-year-old white German model,...  unreliable   
4  A cancer diagnosis never comes at the right ti...  unreliable   
5  In the YouTube video below, you’re going to se...  unreliable   
6  Once a year, the grocery stores start stocking...  unreliable   
7  Starbucks can be a great place to get luxuriou...  unreliable   
8  Dolling out corporal punishment is a controver...  unreliable   
9  He always told her that he liked it rough. But...  unreliable   

                           url_hash  
0  00d76dc2a2b9c02526ab6aba99cea68d  
1  022b655e21ca0d38a1fc33f2a37165a8  
2  0405cdbffb7b0f0aef94eafaba0863ff  
3  09f269f0d217228587029f2012f875c9  
4  0c739d0432f31e7805c0d62f6eff0849  
5  154032234b0955080af9117e817a7676  
6  1af9a9f5326071c7bc5a6e16e04a75df  
7  201402629241e8945bc127ac0b387dc8  
8  21825e99b66d5a6b55e9e3f8c5bdfe71  
9  24c1a494a889e14ef274e14d31fca550

In [18]:
# BIgQuery count by domain
query_job = client.query(
    """
    SELECT
    DOMAIN_HASH
    , LABEL
    , COUNT(*) AS ARTICLES_COUNT
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    GROUP BY DOMAIN_HASH, LABEL
    ORDER BY ARTICLES_COUNT DESC
    """
)

res_df = query_job.result().to_dataframe()  # Waits for job to complete.

# Tally the domain hash to ensure each domain only has one label
domain_tally_ls = []
duplicate_domain_tally_ls = []
for i, row in res_df.iterrows():
  if row['DOMAIN_HASH'] in domain_tally_ls:
    duplicate_domain_tally_ls += [True]
  else:
    duplicate_domain_tally_ls += [False]
  
  # Add domain hash to the list of domains already reviwed
  domain_tally_ls += [row['DOMAIN_HASH']]

res_df['DOMAIN_HAS_MULTIPLE_LABEL'] = duplicate_domain_tally_ls

res_df

DOMAIN_HASH       LABEL  ARTICLES_COUNT  \
0    domain_8f00b2b61ba335244231d632d390bf8d        fake            7046   
1    domain_9f5bbbcbad4a48edd86162fabbe90b6e   political            5441   
2    domain_7e058b16a00bca2c3620d5147881e34d  conspiracy            2819   
3    domain_98845529bda170844508657ae469197a        bias            1162   
4    domain_2bb13bea21c458421e33179105662bdd   political            1061   
5    domain_bd8db3098e8505e2d80c8b89a2eccdb8   clickbait             813   
6    domain_0d6ca907eef322628f81ae5a38735d44        bias             730   
7    domain_f573aed8d0dcf08ef6ce3efd5f5c48c0   political             729   
8    domain_213f2e7f6968dd534eb43ef113fce3e1   political             671   
9    domain_8fbb80fdc3a337fd46babecb01f203fa     junksci             615   
10   domain_799497bdca60cb2f1522d66166f15274   political             435   
11   domain_bfcde262dcc0a3fe874d75293764fb4c   political             415   
12   domain_69a49c3be87af6ae0c48f5c9b005a21b        None             413   
13   domain_f9f6f6f07772425d1522dd8a8ab25193        None             347   
14   domain_5dfa617b9d6dac8de770c521298c4e53     junksci             294   
15   domain_3cbda5292ac5c0b572a8d764b0e52eec        bias             209   
16   domain_fb4e14ee232e6a6acfda1de4f10a1fe3   political             209   
17   domain_c29df259d72ab43dc7e2c43b1941a1c9   clickbait             178   
18   domain_6a12243e8071c910a53039000309c901    reliable             168   
19   domain_7cc6378eb1ac3bf4455eb536225df80b        bias             160   
20   domain_f9e7e3cf456dda4da1f7ada4ba6ddc8c  conspiracy             153   
21   domain_191e287683c85a13fa54c85c43281769  conspiracy             152   
22   domain_3e2b4bc8d67cf01a6f156413acc03375   clickbait             144   
23   domain_1682643b797f67fe89d09642a50174fe   political             120   
24   domain_b55734eb636ef77462a42b666bc74c5f     unknown             114   
25   domain_714bf72800dbdc780c8b895adc12ae3c  unreliable             108   
26   domain_bfff154f86499943be9f3e65798b5d53   political             107   
27   domain_d061e62c09e8c363752ddfa81c54b3ba        bias              99   
28   domain_a66294ec7d2d0d8e3b22c293a88d14f4   political              97   
29   domain_8776e5a68ce1577dd2809b859c3c6d82   political              90   
30   domain_5014c2313218bc048eaef8f34771d2eb     unknown              84   
31   domain_f03bcedc7a83b5568ff986bfec2c3413  conspiracy              77   
32   domain_9725d0802cf38288710d8bff8f64dcba  unreliable              76   
33   domain_fa702aa239d1e1c37cc9c5d75ab9b4f9   political              75   
34   domain_f22938424edda66ec091ddd069854179  unreliable              74   
35   domain_77b6aa19bdd7ac5f26e593472bd9a7d5        hate              71   
36   domain_6afcd6f29c592f5495589242b12d1c25   political              67   
37   domain_2c6ca1a760cdc4987f4283c9464dc639        None              63   
38   domain_090b7aeff850c5dda848d5b9e75bb16d     unknown              58   
39   domain_f047cbddff341a471b7b05bd5f4ee660       rumor              57   
40   domain_daf12fe2bd3c219af940a4f8891f23f0        bias              57   
41   domain_bc6e2fd29bb66b366afa86a3934609ce  conspiracy              54   
42   domain_dbbb88dd5835637c6c4b2e3c54c2669e   political              50   
43   domain_b04f082dca5817be6d561c1e25ea2c16        None              48   
44   domain_b2f824ff829b40551a0f7e99a53745a0      satire              47   
45   domain_e23a8807f4586e5ab98d2c0d335d998a  conspiracy              45   
46   domain_9c34eb473390528f31c0303bda283c21  unreliable              44   
47   domain_95e62162e78b589ec6362d5a6a19b3cb        bias              43   
48   domain_b0c5e6bffbaa50cb2a380e69d80f0fde        bias              42   
49   domain_c072fb2a857d0288f93a765a423629c1  conspiracy              42   
50   domain_9a0d9e918e815c1424ffff912643042f  unreliable              41   
51   domain_f3408542d52cb5a4c7f9d764ba80713b        bias              41   
52   

In [19]:
# BIgQuery count by label
query_job = client.query(
    """
    SELECT
    LABEL
    , COUNT(*) AS LABEL_COUNT
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    GROUP BY LABEL
    ORDER BY LABEL_COUNT DESC
    """
)

res_df = query_job.result().to_dataframe()

print('Total: {}'.format(res_df['LABEL_COUNT'].sum()))

res_df

Total: 27589


LABEL  LABEL_COUNT
0    political         9776
1         fake         7081
2   conspiracy         3512
3         bias         2793
4    clickbait         1238
5      junksci          993
6         None          990
7   unreliable          437
8      unknown          354
9     reliable          178
10      satire           99
11        hate           81
12       rumor           57

In [20]:
# BIgQuery list of URLs
query_job = client.query(
    """
    SELECT DISTINCT
    URL_HASH
    FROM `detect-fake-news-313201.fake_news_sql.src_fake_news`
    """
)

res_df = query_job.result().to_dataframe()

print('Total: {}'.format(res_df.shape[0]))

res_df.head()

Total: 27589


URL_HASH
0  00d76dc2a2b9c02526ab6aba99cea68d
1  022b655e21ca0d38a1fc33f2a37165a8
2  0405cdbffb7b0f0aef94eafaba0863ff
3  09f269f0d217228587029f2012f875c9
4  0c739d0432f31e7805c0d62f6eff0849

Noted that there were no classification for 990 samples, and further 354 with unknown classifications.



## Profile the data in GraphDB

In [21]:
# Install the wrapper package
# Source: https://github.com/RDFLib/sparqlwrapper
!pip install sparqlwrapper

In [30]:
# Code based on: https://sparqlwrapper.readthedocs.io/en/latest/main.html
from SPARQLWrapper import SPARQLWrapper, JSON

queryString = """
PREFIX aa: <http://www.city.ac.uk/ds/inm363/aaron_altrock#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

select (count(?url_hash) as ?url_count) where {
  ?url_hash rdf:type aa:urlHash .
}
"""


sparql = SPARQLWrapper("http://35.246.120.165:7200/repositories/src_fake_news")
sparql.setReturnFormat(JSON)
sparql.setQuery(queryString)

try :
   res_dct = sparql.query().convert()
   print('OK')

except Exception as e:
   print('ERROR: {}'.format(e))
   

OK


In [31]:
# No. of URL hash in GraphDB
res_dct.get('results').get('bindings')[0].get('url_count').get('value')

'27589'

Therefore noted that the number of news articles as URL hashes were completely uploaded when compared to BigQury count given both have the same number of articles `27598`.

In [32]:
res_dct

{'head': {'vars': ['url_count']},
 'results': {'bindings': [{'url_count': {'datatype': 'http://www.w3.org/2001/XMLSchema#integer',
     'type': 'literal',
     'value': '27589'}}]}}

No. of domain hashes

In [79]:
# Code based on: https://sparqlwrapper.readthedocs.io/en/latest/main.html

queryString = """
PREFIX aa: <http://www.city.ac.uk/ds/inm363/aaron_altrock#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

select ?domain_hash (count(?url_hash) as ?url_count) where {
  ?domain_hash rdf:type aa:domainHash .
  ?url_hash rdf:type aa:urlHash .
  ?url_hash aa:has_domain_hash ?domain_hash .
}
GROUP BY ?domain_hash
ORDER BY ?url_count
"""


sparql = SPARQLWrapper("http://35.246.120.165:7200/repositories/src_fake_news")
sparql.setReturnFormat(JSON)
sparql.setQuery(queryString)

try :
   res_dct = sparql.query().convert()
   print('OK')

except Exception as e:
   print('ERROR: {}'.format(e))
   

OK


In [80]:
res_dct

{'head': {'vars': ['domain_hash', 'url_count']},
 'results': {'bindings': [{'domain_hash': {'type': 'uri',
     'value': 'http://www.city.ac.uk/ds/inm363/aaron_altrock#domain_02cb0e25d8e2a8950c674f2d77e05712'},
    'url_count': {'datatype': 'http://www.w3.org/2001/XMLSchema#integer',
     'type': 'literal',
     'value': '1'}},
   {'domain_hash': {'type': 'uri',
     'value': 'http://www.city.ac.uk/ds/inm363/aaron_altrock#domain_06d7d51d4c0251a3b42ebbfe7a26f5b4'},
    'url_count': {'datatype': 'http://www.w3.org/2001/XMLSchema#integer',
     'type': 'literal',
     'value': '1'}},
   {'domain_hash': {'type': 'uri',
     'value': 'http://www.city.ac.uk/ds/inm363/aaron_altrock#domain_1a482c97f790497c9809e63070e977dc'},
    'url_count': {'datatype': 'http://www.w3.org/2001/XMLSchema#integer',
     'type': 'literal',
     'value': '1'}},
   {'domain_hash': {'type': 'uri',
     'value': 'http://www.city.ac.uk/ds/inm363/aaron_altrock#domain_2898ce92e0f40e6ae0fcf2172a57d962'},
    'url_count'

In [91]:
import re
res_ls = res_dct.get('results').get('bindings')

# Helper func to transform SPARQLWrapper query result output (dict) to Pandas dataframe
def parse_to_dataframe(res_ls):

  # If query result has content then parse to data frame else return None
  if len(res_ls) > 0:
    # Get column names
    col_nm_ls = list(res_ls[0].keys())

    parsed_res_ls = []

    for res_dct in res_ls:
      __res_ls = []
      for k, v in res_dct.items():
        __res_ls += [v.get('value')]

      # __res_ls = [res_dct]
      parsed_res_ls += [__res_ls]
    
    res_df = pd.DataFrame.from_dict(parsed_res_ls)
    res_df.columns = col_nm_ls
    res_df.reset_index(inplace=True, drop=True)

    return res_df

  else:
    return None

domain_url_count_df = parse_to_dataframe(res_ls)

# Remove name space prefix
domain_url_count_df['domain_hash'] = domain_url_count_df['domain_hash'].map(lambda str: str[str.find('#') + 1:])
# Convert url_count to integer
domain_url_count_df['url_count'] = domain_url_count_df['url_count'].map(int)

domain_url_count_df.sort_values(by='url_count', ascending=False, inplace=True)

domain_url_count_df

domain_hash  url_count
202  domain_8f00b2b61ba335244231d632d390bf8d       7046
201  domain_9f5bbbcbad4a48edd86162fabbe90b6e       5441
200  domain_7e058b16a00bca2c3620d5147881e34d       2819
199  domain_98845529bda170844508657ae469197a       1162
198  domain_2bb13bea21c458421e33179105662bdd       1061
197  domain_bd8db3098e8505e2d80c8b89a2eccdb8        813
196  domain_0d6ca907eef322628f81ae5a38735d44        730
195  domain_f573aed8d0dcf08ef6ce3efd5f5c48c0        729
194  domain_213f2e7f6968dd534eb43ef113fce3e1        671
193  domain_8fbb80fdc3a337fd46babecb01f203fa        615
192  domain_799497bdca60cb2f1522d66166f15274        435
191  domain_bfcde262dcc0a3fe874d75293764fb4c        415
190  domain_69a49c3be87af6ae0c48f5c9b005a21b        413
189  domain_f9f6f6f07772425d1522dd8a8ab25193        347
188  domain_5dfa617b9d6dac8de770c521298c4e53        294
187  domain_fb4e14ee232e6a6acfda1de4f10a1fe3        209
186  domain_3cbda5292ac5c0b572a8d764b0e52eec        209
185  domain_c29df259d72ab43dc7e2c43b1941a1c9        178
184  domain_6a12243e8071c910a53039000309c901        168
183  domain_7cc6378eb1ac3bf4455eb536225df80b        160
182  domain_f9e7e3cf456dda4da1f7ada4ba6ddc8c        153
181  domain_191e287683c85a13fa54c85c43281769        152
180  domain_3e2b4bc8d67cf01a6f156413acc03375        144
179  domain_1682643b797f67fe89d09642a50174fe        120
178  domain_b55734eb636ef77462a42b666bc74c5f        114
177  domain_714bf72800dbdc780c8b895adc12ae3c        108
176  domain_bfff154f86499943be9f3e65798b5d53        107
175  domain_d061e62c09e8c363752ddfa81c54b3ba         99
174  domain_a66294ec7d2d0d8e3b22c293a88d14f4         97
173  domain_8776e5a68ce1577dd2809b859c3c6d82         90
172  domain_5014c2313218bc048eaef8f34771d2eb         84
171  domain_f03bcedc7a83b5568ff986bfec2c3413         77
170  domain_9725d0802cf38288710d8bff8f64dcba         76
169  domain_fa702aa239d1e1c37cc9c5d75ab9b4f9         75
168  domain_f22938424edda66ec091ddd069854179         74
167  domain_77b6aa19bdd7ac5f26e593472bd9a7d5         71
166  domain_6afcd6f29c592f5495589242b12d1c25         67
165  domain_2c6ca1a760cdc4987f4283c9464dc639         63
164  domain_090b7aeff850c5dda848d5b9e75bb16d         58
163  domain_f047cbddff341a471b7b05bd5f4ee660         57
162  domain_daf12fe2bd3c219af940a4f8891f23f0         57
161  domain_bc6e2fd29bb66b366afa86a3934609ce         54
160  domain_dbbb88dd5835637c6c4b2e3c54c2669e         50
159  domain_b04f082dca5817be6d561c1e25ea2c16         48
158  domain_b2f824ff829b40551a0f7e99a53745a0         47
157  domain_e23a8807f4586e5ab98d2c0d335d998a         45
156  domain_9c34eb473390528f31c0303bda283c21         44
155  domain_95e62162e78b589ec6362d5a6a19b3cb         43
154  domain_c072fb2a857d0288f93a765a423629c1         42
153  domain_b0c5e6bffbaa50cb2a380e69d80f0fde         42
152  domain_f3408542d52cb5a4c7f9d764ba80713b         41
151  domain_9a0d9e918e815c1424ffff912643042f         41
150  domain_e1e8ae12f228f5344ff99f14eb6926e8         33
149  domain_dc7dbd0a07a4d8dcf91c1d6de72611ea         33
148  domain_1d3355bc85a66dfd13a09f3412c7f085         33
147  domain_f446297823e3381ccbdcadd1033349f8         32
146  domain_ae885395fb95bb58ee77f8ae3925d64c         32
145  domain_1c74052b1eb7f1b234c3c19714e65a53         31
144  domain_f9a3693da598b15b2c9235cb8c2c3e74         30
143  domain_2e59c440f7b5b4a04a68f652f5b5333b         28
141  domain_670f9a1790abd5f611073d167d0b8181         27
142  domain_d9b91044b82468a930dbba5a1be97efe         27
140  domain_e405e81e7af8517d01f31f4148fc71f8         26
139  domain_e34675b4f28a906f84eeae9b500db072         26
138  domain_0691d86f4f2449106e6b812635ae6dc2         24
137  domain_e0ed6141030ad9697efb6c0fbeaaf39a         23
136  domain_7e6e1c6f7cf959d4b9bb48a6343ae80a         23
135  domain_902091121d014e622d75cbc7366f2ee6         22
134  domain_ef1bc2898cde8803366bc549a5db7810         21
133  domain_bf1876a5bd135bede0c8d6dceee56287         21
131  domain_77bb545e808a339883519ec87b37568b         20
1